File: modules/creepy-android-settings/src/CreepyAndroidSettingsModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import {
  assertBrightnessInRange,
  assertPercentInRange,
  assertScreenTimeoutNonNegative,
} from './conversion';
import type {
  AndroidSettingsCapabilities,
  CreepyAndroidSettingsAPI,
  CreepyAndroidSettingsEvents,
  SettingsNamespace,
  SettingsPanel,
  SettingsScreen,
} from './CreepyAndroidSettings.types';

declare class CreepyAndroidSettingsNativeModule extends NativeModule<CreepyAndroidSettingsEvents> {
  getCapabilities(): AndroidSettingsCapabilities;

  getSetting(namespace: string, key: string): string | null;
  getSettingInt(namespace: string, key: string, defaultValue: number): number;

  getSystemInt(key: string, defaultValue: number): number;
  getSystemString(key: string): string | null;
  getSystemFloat(key: string, defaultValue: number): number;
  getSystemLong(key: string, defaultValue: number): number;

  setSystemInt(key: string, value: number): boolean;
  setSystemString(key: string, value: string): boolean;
  setSystemFloat(key: string, value: number): boolean;
  setSystemLong(key: string, value: number): boolean;

  getSecureInt(key: string, defaultValue: number): number;
  getSecureString(key: string): string | null;
  getSecureFloat(key: string, defaultValue: number): number;
  getSecureLong(key: string, defaultValue: number): number;

  getGlobalInt(key: string, defaultValue: number): number;
  getGlobalString(key: string): string | null;
  getGlobalFloat(key: string, defaultValue: number): number;
  getGlobalLong(key: string, defaultValue: number): number;

  canWriteSystemSettings(): boolean;
  requestWriteSystemSettingsPermission(): Promise<boolean>;
  canDrawOverlays(): boolean;
  requestOverlayPermission(): Promise<boolean>;

  getScreenBrightness(): number;
  setScreenBrightness(value: number): boolean;
  getScreenBrightnessPercent(): number;
  setScreenBrightnessPercent(percent: number): boolean;
  getScreenTimeout(): number;
  setScreenTimeout(milliseconds: number): boolean;
  getAutoRotate(): boolean;
  setAutoRotate(enabled: boolean): boolean;

  canOpenSettings(screen: string): boolean;
  openSettings(screen: string): Promise<boolean>;

  isSettingsPanelSupported(panel: string): boolean;
  openPanel(panel: string): Promise<boolean>;

  watchSetting(namespace: string, key: string): string;
  unwatchSetting(subscriptionId: string): void;
  unwatchAllSettings(): void;
}

const nativeModule =
  requireNativeModule<CreepyAndroidSettingsNativeModule>('CreepyAndroidSettings');

export const CreepyAndroidSettings: CreepyAndroidSettingsAPI = {
  getCapabilities: () => nativeModule.getCapabilities(),

  getSetting: (namespace: SettingsNamespace, key: string) => nativeModule.getSetting(namespace, key),
  getSettingInt: (namespace: SettingsNamespace, key: string, defaultValue?: number) =>
    nativeModule.getSettingInt(namespace, key, defaultValue ?? 0),

  getSystemInt: (key: string, defaultValue?: number) => nativeModule.getSystemInt(key, defaultValue ?? 0),
  getSystemString: (key: string) => nativeModule.getSystemString(key),
  getSystemFloat: (key: string, defaultValue?: number) => nativeModule.getSystemFloat(key, defaultValue ?? 0),
  getSystemLong: (key: string, defaultValue?: number) => nativeModule.getSystemLong(key, defaultValue ?? 0),

  setSystemInt: (key: string, value: number) => nativeModule.setSystemInt(key, value),
  setSystemString: (key: string, value: string) => nativeModule.setSystemString(key, value),
  setSystemFloat: (key: string, value: number) => nativeModule.setSystemFloat(key, value),
  setSystemLong: (key: string, value: number) => nativeModule.setSystemLong(key, value),

  getSecureInt: (key: string, defaultValue?: number) => nativeModule.getSecureInt(key, defaultValue ?? 0),
  getSecureString: (key: string) => nativeModule.getSecureString(key),
  getSecureFloat: (key: string, defaultValue?: number) => nativeModule.getSecureFloat(key, defaultValue ?? 0),
  getSecureLong: (key: string, defaultValue?: number) => nativeModule.getSecureLong(key, defaultValue ?? 0),

  getGlobalInt: (key: string, defaultValue?: number) => nativeModule.getGlobalInt(key, defaultValue ?? 0),
  getGlobalString: (key: string) => nativeModule.getGlobalString(key),
  getGlobalFloat: (key: string, defaultValue?: number) => nativeModule.getGlobalFloat(key, defaultValue ?? 0),
  getGlobalLong: (key: string, defaultValue?: number) => nativeModule.getGlobalLong(key, defaultValue ?? 0),

  canWriteSystemSettings: () => nativeModule.canWriteSystemSettings(),
  requestWriteSystemSettingsPermission: () => nativeModule.requestWriteSystemSettingsPermission(),
  canDrawOverlays: () => nativeModule.canDrawOverlays(),
  requestOverlayPermission: () => nativeModule.requestOverlayPermission(),

  getScreenBrightness: () => nativeModule.getScreenBrightness(),
  setScreenBrightness: (value: number) => {
    assertBrightnessInRange(value);
    return nativeModule.setScreenBrightness(value);
  },
  getScreenBrightnessPercent: () => nativeModule.getScreenBrightnessPercent(),
  setScreenBrightnessPercent: (percent: number) => {
    assertPercentInRange(percent);
    return nativeModule.setScreenBrightnessPercent(percent);
  },
  getScreenTimeout: () => nativeModule.getScreenTimeout(),
  setScreenTimeout: (milliseconds: number) => {
    assertScreenTimeoutNonNegative(milliseconds);
    return nativeModule.setScreenTimeout(milliseconds);
  },
  getAutoRotate: () => nativeModule.getAutoRotate(),
  setAutoRotate: (enabled: boolean) => nativeModule.setAutoRotate(enabled),

  canOpenSettings: (screen: SettingsScreen) => nativeModule.canOpenSettings(screen),
  openSettings: (screen: SettingsScreen) => nativeModule.openSettings(screen),

  isSettingsPanelSupported: (panel: SettingsPanel) => nativeModule.isSettingsPanelSupported(panel),
  openPanel: (panel: SettingsPanel) => nativeModule.openPanel(panel),

  watchSetting: (namespace: SettingsNamespace, key: string) => nativeModule.watchSetting(namespace, key),
  unwatchSetting: (subscriptionId: string) => nativeModule.unwatchSetting(subscriptionId),
  unwatchAllSettings: () => nativeModule.unwatchAllSettings(),

  addListener: (eventName, listener) => nativeModule.addListener(eventName, listener),
};

File: modules/creepy-android-settings/android/src/main/java/expo/modules/creepyandroidsettings/CreepyAndroidSettingsModule.kt
package expo.modules.creepyandroidsettings

import android.content.Context
import android.os.Build
import android.provider.Settings
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo Modules API surface for the Creepy.IM Android settings bridge.
 *
 * This module only wires the Kotlin helpers together; all responsibilities are
 * split across [SettingsReader], [SettingsWriter], [SettingsNavigator] and
 * [SettingsObserverManager].
 */
class CreepyAndroidSettingsModule : Module() {

    private val context: Context
        get() = appContext.reactContext ?: throw AndroidContextUnavailableException()

    private val reader: SettingsReader by lazy { SettingsReader(context.contentResolver) }
    private val writer: SettingsWriter by lazy { SettingsWriter(context) }
    private val navigator: SettingsNavigator by lazy { SettingsNavigator(context) }

    private var observerManager: SettingsObserverManager? = null

    private fun requireObserverManager(): SettingsObserverManager {
        return observerManager ?: SettingsObserverManager(context.contentResolver) { id, namespace, key, value, timestamp ->
            sendEvent(
                "onSettingChanged",
                mapOf(
                    "subscriptionId" to id,
                    "namespace" to namespace.value,
                    "key" to key,
                    "value" to value,
                    "timestamp" to timestamp,
                ),
            )
        }.also { observerManager = it }
    }

    override fun definition() = ModuleDefinition {
        Name("CreepyAndroidSettings")

        Events("onSettingChanged")

        OnDestroy {
            observerManager?.unwatchAll()
        }

        // region Capabilities

        Function("getCapabilities") {
            val supportedScreens = linkedMapOf<String, Boolean>()
            for (screen in SettingsNavigator.SCREENS) {
                supportedScreens[screen] = navigator.canOpenScreen(screen)
            }

            val capabilities: Map<String, Any?> = linkedMapOf(
                "platform" to "android",
                "apiLevel" to Build.VERSION.SDK_INT,
                "manufacturer" to Build.MANUFACTURER,
                "model" to Build.MODEL,
                "canWriteSystemSettings" to Settings.System.canWrite(context),
                "canDrawOverlays" to Settings.canDrawOverlays(context),
                "settingsPanelsSupported" to (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q),
                "supportedScreens" to supportedScreens,
            )
            capabilities
        }

        // endregion

        // region Generic reads

        Function("getSetting") { namespace: String, key: String ->
            requireKey(key)
            reader.getString(SettingsNamespace.from(namespace), key)
        }

        Function("getSettingInt") { namespace: String, key: String, defaultValue: Int ->
            requireKey(key)
            reader.getInt(SettingsNamespace.from(namespace), key, defaultValue)
        }

        // endregion

        // region Settings.System reads

        Function("getSystemInt") { key: String, defaultValue: Int ->
            requireKey(key)
            reader.getInt(SettingsNamespace.SYSTEM, key, defaultValue)
        }

        Function("getSystemString") { key: String ->
            requireKey(key)
            reader.getString(SettingsNamespace.SYSTEM, key)
        }

        Function("getSystemFloat") { key: String, defaultValue: Float ->
            requireKey(key)
            reader.getFloat(SettingsNamespace.SYSTEM, key, defaultValue)
        }

        Function("getSystemLong") { key: String, defaultValue: Long ->
            requireKey(key)
            reader.getLong(SettingsNamespace.SYSTEM, key, defaultValue)
        }

        // endregion

        // region Settings.Secure reads

        Function("getSecureInt") { key: String, defaultValue: Int ->
            requireKey(key)
            reader.getInt(SettingsNamespace.SECURE, key, defaultValue)
        }

        Function("getSecureString") { key: String ->
            requireKey(key)
            reader.getString(SettingsNamespace.SECURE, key)
        }

        Function("getSecureFloat") { key: String, defaultValue: Float ->
            requireKey(key)
            reader.getFloat(SettingsNamespace.SECURE, key, defaultValue)
        }

        Function("getSecureLong") { key: String, defaultValue: Long ->
            requireKey(key)
            reader.getLong(SettingsNamespace.SECURE, key, defaultValue)
        }

        // endregion

        // region Settings.Global reads

        Function("getGlobalInt") { key: String, defaultValue: Int ->
            requireKey(key)
            reader.getInt(SettingsNamespace.GLOBAL, key, defaultValue)
        }

        Function("getGlobalString") { key: String ->
            requireKey(key)
            reader.getString(SettingsNamespace.GLOBAL, key)
        }

        Function("getGlobalFloat") { key: String, defaultValue: Float ->
            requireKey(key)
            reader.getFloat(SettingsNamespace.GLOBAL, key, defaultValue)
        }

        Function("getGlobalLong") { key: String, defaultValue: Long ->
            requireKey(key)
            reader.getLong(SettingsNamespace.GLOBAL, key, defaultValue)
        }

        // endregion

        // region Settings.System writes (LOW LEVEL API — use the high-level helpers instead)

        Function("setSystemInt") { key: String, value: Int ->
            requireKey(key)
            writer.putInt(key, value)
        }

        Function("setSystemString") { key: String, value: String ->
            requireKey(key)
            writer.putString(key, value)
        }

        Function("setSystemFloat") { key: String, value: Float ->
            requireKey(key)
            writer.putFloat(key, value)
        }

        Function("setSystemLong") { key: String, value: Long ->
            requireKey(key)
            writer.putLong(key, value)
        }

        // endregion

        // region Special access

        Function("canWriteSystemSettings") {
            Settings.System.canWrite(context)
        }

        AsyncFunction("requestWriteSystemSettingsPermission") {
            if (navigator.canOpenWriteSettings()) {
                navigator.openWriteSettings()
            }
            Settings.System.canWrite(context)
        }

        Function("canDrawOverlays") {
            Settings.canDrawOverlays(context)
        }

        AsyncFunction("requestOverlayPermission") {
            if (navigator.canOpenOverlay()) {
                navigator.openOverlay()
            }
            Settings.canDrawOverlays(context)
        }

        // endregion

        // region High-level settings

        Function("getScreenBrightness") {
            reader.getInt(SettingsNamespace.SYSTEM, Settings.System.SCREEN_BRIGHTNESS, 0)
        }

        Function("setScreenBrightness") { value: Int ->
            requireInRange(value, 0, 255, "brightness")
            writer.putInt(Settings.System.SCREEN_BRIGHTNESS, value)
        }

        Function("getScreenBrightnessPercent") {
            val brightness = reader.getInt(SettingsNamespace.SYSTEM, Settings.System.SCREEN_BRIGHTNESS, 0)
            percentFromBrightness(brightness)
        }

        Function("setScreenBrightnessPercent") { percent: Int ->
            requireInRange(percent, 0, 100, "percent")
            writer.putInt(Settings.System.SCREEN_BRIGHTNESS, brightnessFromPercent(percent))
        }

        Function("getScreenTimeout") {
            reader.getInt(SettingsNamespace.SYSTEM, Settings.System.SCREEN_OFF_TIMEOUT, 0)
        }

        Function("setScreenTimeout") { milliseconds: Int ->
            requireNonNegative(milliseconds, "milliseconds")
            writer.putInt(Settings.System.SCREEN_OFF_TIMEOUT, milliseconds)
        }

        Function("getAutoRotate") {
            reader.getInt(SettingsNamespace.SYSTEM, Settings.System.ACCELEROMETER_ROTATION, 0) == 1
        }

        Function("setAutoRotate") { enabled: Boolean ->
            writer.putInt(Settings.System.ACCELEROMETER_ROTATION, if (enabled) 1 else 0)
        }

        // endregion

        // region Navigation

        Function("canOpenSettings") { screen: String ->
            navigator.canOpenScreen(screen)
        }

        AsyncFunction("openSettings") { screen: String ->
            navigator.openScreen(screen)
        }

        Function("isSettingsPanelSupported") { panel: String ->
            navigator.isPanelSupported(panel)
        }

        AsyncFunction("openPanel") { panel: String ->
            navigator.openPanel(panel)
        }

        // endregion

        // region Observer

        Function("watchSetting") { namespace: String, key: String ->
            requireKey(key)
            requireObserverManager().watch(SettingsNamespace.from(namespace), key)
        }

        Function("unwatchSetting") { subscriptionId: String ->
            requireObserverManager().unwatch(subscriptionId)
        }

        Function("unwatchAllSettings") {
            observerManager?.unwatchAll()
            Unit
        }

        // endregion
    }

    private fun requireKey(key: String) {
        if (key.isBlank()) {
            throw InvalidArgumentException("Settings key must not be empty.")
        }
    }

    private fun requireInRange(value: Int, min: Int, max: Int, name: String) {
        if (value < min || value > max) {
            throw InvalidArgumentException("$name must be within $min..$max, received $value.")
        }
    }

    private fun requireNonNegative(value: Int, name: String) {
        if (value < 0) {
            throw InvalidArgumentException("$name must be >= 0, received $value.")
        }
    }

    private fun brightnessFromPercent(percent: Int): Int =
        Math.round(percent / 100.0 * 255.0).toInt().coerceIn(0, 255)

    private fun percentFromBrightness(brightness: Int): Int =
        Math.round(brightness / 255.0 * 100.0).toInt().coerceIn(0, 100)
}
